### ISSUE TITLE
How to debug my performance issue?

### ISSUE THREAD

**Issue by huangya90** [meyerd/n2n#19]
Hi, I am encountering a performance issue. After the tunnel was ｅｓｔａｂlished, I tried playing the high-definition video via the tunnel, but it was not very smooth. Then I used iperf to test the performance, Here is data what I got. Rate about 19 Mbits/sec is what I expect. However, according to the test result,  I need to use multi sessions(two or above) to get such a rate. Only about 12 Mbits/sec for 1 session. The difference of 7Mbits/sec is very important to  play the high-definition video. Any suggestion? Thanks for your time and  great job!

XXXX% iperf -c 192.168.6.64     

Client connecting to 192.168.6.64, TCP port 5001
TCP window size: 45.0 KByte (default)

[  3] local 192.168.6.2 port 37024 connected with 192.168.6.64 port 5001
[ ID] Interval       Transfer     Bandwidth
[  3]  0.0-10.2 sec  15.2 MBytes  **12.6 Mbits/sec**
XXXX% iperf -c 192.168.6.64 

Client connecting to 192.168.6.64, TCP port 5001
TCP window size: 45.0 KByte (default)

[  3] local 192.168.6.2 port 37026 connected with 192.168.6.64 port 5001
[ ID] Interval       Transfer     Bandwidth
[  3]  0.0-10.1 sec  13.0 MBytes  **10.8 Mbits/sec**
XXXX% iperf -c 192.168.6.64 -P 2

Client connecting to 192.168.6.64, TCP port 5001
TCP window size: 45.0 KByte (default)

[  4] local 192.168.6.2 port 37034 connected with 192.168.6.64 port 5001
[  3] local 192.168.6.2 port 37032 connected with 192.168.6.64 port 5001
[ ID] Interval       Transfer     Bandwidth
[  4]  0.0-10.1 sec  11.4 MBytes  9.40 Mbits/sec
[  3]  0.0-10.2 sec  12.4 MBytes  10.2 Mbits/sec
[SUM]  0.0-10.2 sec  23.8 MBytes  **19.6 Mbits/sec**
XXXX% iperf -c 192.168.6.64 -P 2

Client connecting to 192.168.6.64, TCP port 5001
TCP window size: 45.0 KByte (default)

[  4] local 192.168.6.2 port 37036 connected with 192.168.6.64 port 5001
[  3] local 192.168.6.2 port 37038 connected with 192.168.6.64 port 5001
[ ID] Interval       Transfer     Bandwidth
[  4]  0.0-10.2 sec  12.1 MBytes  10.0 Mbits/sec
[  3]  0.0-10.2 sec  11.5 MBytes  9.49 Mbits/sec
[SUM]  0.0-10.2 sec  23.6 MBytes  **19.5 Mbits/sec**

**Comment by meyerd** [comment]
I guess that could be either a performance issue on the supernode or on the edge. Could you look into the cpu usage of either? Also for debugging which function takes the most time I would recommend using the default C profiling tools such as gprof as I think the transfer rate is CPU bound due to some poor implementations in packet handling.

**Comment by huangya90** [comment]
@meyerd.  In my case, CPU is not the bottleneck. My CPUs are 2 cores and 4 threads on two edges. It's  about 50% usage of one thread. I found the direct cause is related with wireless. One edge is connected with a wireless router. If I let the edge connected with wire, the issue is gone. I can get 20Mbit/sec. However, I am sure the wireless link quality is good because I tested the speed via www.speedtest.net and got expected speed. I will do more test to find out the root cause. Any more suggestions are welcome.

**Comment by huangya90** [comment]
@meyerd , I have found the root cause. Once I played with my router and set the MTU size to 1300, which causes the fragmentation. Now I set the MTU size to 1492 in PPPoE mode, it's normal as below. I am not sure why the wire is OK. Maybe because there is hardware NAT in my router and the procedures between wire and wireless are different. Close the issue.

XXXX% iperf -c 192.168.6.64 -t 15

Client connecting to 192.168.6.64, TCP port 5001
TCP window size: 45.0 KByte (default)

[  3] local 192.168.6.2 port 39104 connected with 192.168.6.64 port 5001
[ ID] Interval       Transfer     Bandwidth
[  3]  0.0-15.1 sec  41.0 MBytes  22.8 Mbits/sec
